harness/determined/estimator/_reducer.py

```diff
@@ -117,38 +117,102 @@ def cross_slot_reduce(self, per_slot_metrics: List[List[Any]]) -> Any:
         return self.reduce_fn(flat_metrics)
 
 
-def _distributed_metric(
-    context: estimator.EstimatorExperimentalContext,
-    metric: Any,
-    reducer: MetricReducer,
-    numpy_dtype: Any,
-) -> Tuple[tf.Operation, tf.Operation]:
+def _deconstruct_metric(metric: Any):
+    """Break down lists and dictionaries into something that can be passed through at graph."""
+    if isinstance(metric, (tf.Tensor, tf.Operation)):
+        return [metric], (tf.Tensor, None)
+
+    if isinstance(metric, (list, tuple)):
+        for m in metric:
+            if not isinstance(m, (tf.Tensor, tf.Operation)):
+                raise TypeError(
+                    "list-type metric parameters must be a flat list of tf.Tensors but found "
+                    f"key of type {type(k)}"
+                )
+        return metric, (list, None)
+
+    if isinstance(metric, dict):
+        for k, v in metric.items():
+            if not isinstance(k, str):
+                raise TypeError(
+                    "dict-type metric parameters must be a flat list mapping strings to "
+                    f"tf.Tensors but found key of type {type(k)}"
+                )
+            if not isinstance(v, (tf.Tensor, tf.Operation)):
+                raise TypeError(
+                    "dict-type metric parameters must be a flat list mapping strings to "
+                    f"tf.Tensors but found value of type {type(v)}"
+                )
+        keys, args = zip(*metric.items())
+        return args, (dict, keys)
+
+    else:
+        # Try to convert the arbitrary input to a constant Tensor.
+        try:
+            const_metric = tf.compat.v1.constant(metric)
+        except TypeError:
+            raise TypeError(
+                "metric parameter must be a tf.Tensor, a list of tf.Tensors, "
+                f"or a dict mapping strings to tf.Tensors, not {type(metric)}"
+            )
+        return const_metric, (tf.Tensor, None)
+
+
+def _reconstruct_metric(args: List, reconstruct_info: Tuple[type, List]):
+    """Reconstruct lists or dictionaries after passing them through the graph."""
+    metric_type, update_keys = reconstruct_info
+    if metric_type == tf.Tensor:
+        return args[0]
+    if metric_type == list:
+        return args
+    if metric_type == dict:
+        return {k: v for k, v in zip(update_keys, args)}
+    raise AssertionError(f"invalid metric_type: {metric_type}")
+
+
+class _DistributedMetricMaker:
```

Comment:
Personally I don't find the class version better than the function version with inline functions and the one `nonlocal` statement.  I would prefer to write a comment explaining what the `nonlocal` statement is for than to use a class here.

Reply:
@aaron276h says he likes the new one a *lot* better, and I only like the old way a *little* better.  So we'll stick with this.